firebase-storage/src/main/java/com/google/firebase/storage/StorageReference.java

```diff
@@ -563,14 +566,119 @@ public StreamDownloadTask getStream(@NonNull StreamDownloadTask.StreamProcessor
+   *     StorageReference.
+   */
+  @NonNull
+  @PublicApi
+  public Task<ListResult> list(int maxResults) {
+    Preconditions.checkArgument(maxResults > 0, "maxResults must be greater than zero");
+    Preconditions.checkArgument(maxResults <= 1000, "maxResults must be less than 1000");
+    TaskCompletionSource<ListResult> pendingResult = new TaskCompletionSource<>();
+    StorageTaskScheduler.getInstance()
+        .scheduleCommand(new ListTask(this, maxResults, /* pageToken= */ null, pendingResult));
+    return pendingResult.getTask();
+  }
+
+  /**
+   * Resumes a previous call to {@link #list(int)}, starting after a pagination token. Returns the
+   * next set of items (files) and prefixes (folders) under this StorageReference.
+   *
+   * <p>"/" is treated as a path delimiter. Firebase Storage does not support invalid object paths
+   * that end with "/" or contain two consecutive "//". All invalid objects in GCS will be filtered.
+   *
+   * @param maxResults The maximum number of results to return in a single page. Must be greater
+   *     than 0 and at most 1000.
+   * @param pageToken A page token from a previous call to list.
+   * @return A a {@link Task} that returns the next items and prefixes under the current
+   *     StorageReference.
+   */
+  @NonNull
+  @PublicApi
+  public Task<ListResult> list(int maxResults, @NonNull String pageToken) {
+    Preconditions.checkArgument(maxResults > 0, "maxResults must be greater than zero");
+    Preconditions.checkArgument(maxResults <= 1000, "maxResults must be less than 1000");
+    Preconditions.checkArgument(
+        pageToken != null, "pageToken must be non-null to resume a previous list() operation");
+    TaskCompletionSource<ListResult> pendingResult = new TaskCompletionSource<>();
+    StorageTaskScheduler.getInstance()
+        .scheduleCommand(new ListTask(this, maxResults, pageToken, pendingResult));
+    return pendingResult.getTask();
+  }
+
+  /**
+   * List all items (files) and prefixes (folders) under this StorageReference.
+   *
+   * <p>This is a helper method for calling list() repeatedly until there are no more results. The
+   * default pagination size is 1000. Consistency of the result is not guaranteed if objects are
+   * inserted or removed while this operation is executing.
+   *
+   * @throws OutOfMemoryError If there are too many items at this location.
+   * @return A {@link Task} that returns all items and prefixes under the current StorageReference.
+   */
+  @NonNull
+  @PublicApi
+  public Task<ListResult> listAll() {
+    TaskCompletionSource<ListResult> pendingResult = new TaskCompletionSource<>();
+
+    List<ListResult> results = new ArrayList<>();
+    int[] prefixesCount = new int[] {0};
+    int[] itemsCount = new int[] {0};
+
+    Executor executor = StorageTaskScheduler.getInstance().getCommandPoolExecutor();
+    Task<ListResult> list = list(ListTask.DEFAULT_PAGE_SIZE);
```

Comment:
One caveat that I should explain here.
1. When you pass in `maxResults` to Firebase Storage Backend, we run a loop to make sure either max results are reached or there is no more results.
2. When you do not pass in `maxResults`, we always call GCS once and return the data.

Both of them would filter invalid objects. If there is one invalid object (very common case if you create a directory in console), 
1. In the first case, we call GCS twice to fill the spot for that invalid object.
2. In the second case, we return `999=1000-1` elements.

So in practice, not passing in a `maxResults` is a lot more effecient for listAll.

Reply:
Thanks for the explanation. Guess it makes sense for the SDK and the backend team to collaborate :) I made maxResults optional.